### PR TITLE
[3.4] allow downgrade from 3.5

### PR DESCRIPTION
**Pull request by siyuanfoundation**
Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

Part of https://github.com/etcd-io/etcd/issues/15878#issuecomment-1910886682
Tested downgrade from 3.5 cluster locally:

1. Start 3.5 cluster
```
bin/etcd --name infra1 --listen-client-urls http://127.0.0.1:2379 --advertise-client-urls http://127.0.0.1:2379 --listen-peer-urls http://127.0.0.1:12380 --initial-advertise-peer-urls http://127.0.0.1:12380 --initial-cluster-token etcd-cluster-1 --initial-cluster 'infra1=http://127.0.0.1:12380,infra2=http://127.0.0.1:22380,infra3=http://127.0.0.1:32380' --initial-cluster-state new --enable-pprof --logger=zap --log-outputs=stderr

bin/etcd --name infra2 --listen-client-urls http://127.0.0.1:22379 --advertise-client-urls http://127.0.0.1:22379 --listen-peer-urls http://127.0.0.1:22380 --initial-advertise-peer-urls http://127.0.0.1:22380 --initial-cluster-token etcd-cluster-1 --initial-cluster 'infra1=http://127.0.0.1:12380,infra2=http://127.0.0.1:22380,infra3=http://127.0.0.1:32380' --initial-cluster-state new --enable-pprof --logger=zap --log-outputs=stderr

bin/etcd --name infra3 --listen-client-urls http://127.0.0.1:32379 --advertise-client-urls http://127.0.0.1:32379 --listen-peer-urls http://127.0.0.1:32380 --initial-advertise-peer-urls http://127.0.0.1:32380 --initial-cluster-token etcd-cluster-1 --initial-cluster 'infra1=http://127.0.0.1:12380,infra2=http://127.0.0.1:22380,infra3=http://127.0.0.1:32380' --initial-cluster-state new --enable-pprof --logger=zap --log-outputs=stderr
```
2. start putting data 
```
go run tools/benchmark/main.go put --total=10000000 --key-size=64 --val-size=512`

bin/etcdctl --endpoints=http://127.0.0.1:2379,http://127.0.0.1:22379,http://127.0.0.1:32379 endpoint status -w table
bin/etcdctl auth status
```
3. kill `infra3`
4. restart `infra3` with 3.4 binary
```
../etcd-3.4/bin/etcd --name infra3 --listen-client-urls http://127.0.0.1:32379 --advertise-client-urls http://127.0.0.1:32379 --listen-peer-urls http://127.0.0.1:32380 --initial-advertise-peer-urls http://127.0.0.1:32380 --initial-cluster 'infra1=http://127.0.0.1:12380,infra2=http://127.0.0.1:22380,infra3=http://127.0.0.1:32380' --initial-cluster-state existing --enable-pprof --logger=zap --log-outputs=stderr --data-dir infra3.etcd --accept-next-storage-version

bin/etcdctl auth status
```
5. kill and restart `infra2` with 3.4 binary
```
../etcd-3.4/bin/etcd --name infra2 --listen-client-urls http://127.0.0.1:22379 --advertise-client-urls http://127.0.0.1:22379 --listen-peer-urls http://127.0.0.1:22380 --initial-advertise-peer-urls http://127.0.0.1:22380 --initial-cluster-token etcd-cluster-1 --initial-cluster 'infra1=http://127.0.0.1:12380,infra2=http://127.0.0.1:22380,infra3=http://127.0.0.1:32380' --initial-cluster-state existing --enable-pprof --logger=zap --log-outputs=stderr --data-dir infra2.etcd --accept-next-storage-version
```
6. kill and restart `infra1` with 3.4 binary
```
../etcd-3.4/bin/etcd --name infra1 --listen-client-urls http://127.0.0.1:2379 --advertise-client-urls http://127.0.0.1:2379 --listen-peer-urls http://127.0.0.1:12380 --initial-advertise-peer-urls http://127.0.0.1:12380 --initial-cluster-token etcd-cluster-1 --initial-cluster 'infra1=http://127.0.0.1:12380,infra2=http://127.0.0.1:22380,infra3=http://127.0.0.1:32380' --initial-cluster-state existing --enable-pprof --logger=zap --log-outputs=stderr --data-dir infra1.etcd --accept-next-storage-version
```

### CODE REVIEW

**Comment by siyuanfoundation** [comment]
cc @serathius @ahrtr

**Comment by serathius** [comment]
Not sure how this PR refers to dev work described in https://docs.google.com/document/d/1siSo-tA_xLRo5li5wK8KqWsOiq34PIgtmuSDcyxxwsY/edit?usp=sharing

In document you described that downgrade will gated behind "--allow-downgrade" flag and will include implementation of AuthStatusRequest handling.

File: mvcc/kvstore.go
Comment (serathius) [inline mvcc/kvstore.go:382]: You picked a pretty good place to add logic for handling db file. I think that I even missed in the original downgrade design that only migrates storage when bootstrapping etcd. Added https://github.com/etcd-io/etcd/issues/17361 to improve it.

However the validation you added would not work. What happens if during downgrade v3.4 member gets a snapshot from leader on v3.5? We need to reimplement parts of migrate code to detect storage version and downgrade it to v3.4, so we remove "term" and "confState" if encountered.

File: etcdserver/api/membership/cluster.go
Comment (serathius) [inline etcdserver/api/membership/cluster.go:798]: Based on design I expected this will be version gated.

**Comment by siyuanfoundation** [comment]
> Not sure how this PR refers to dev work described in https://docs.google.com/document/d/1siSo-tA_xLRo5li5wK8KqWsOiq34PIgtmuSDcyxxwsY/edit?usp=sharing
> 
> In document you described that downgrade will gated behind "--allow-downgrade" flag and will include implementation of AuthStatusRequest handling.

With the `etcdutil migrate` and online schema check, "--allow-downgrade" flag is kind of redundant. 
But I see your point about snapshot.

Added "--allow-downgrade" flag and implementation of AuthStatusRequest handling.

File: mvcc/kvstore.go
Comment (siyuanfoundation) [reply mvcc/kvstore.go]: done

**Comment by siyuanfoundation** [comment]
cc @lavacat

File: embed/config.go
Comment (serathius) [inline embed/config.go:367]: Users might not be aware of storage version, which was also introduced in v3.6.

Naming is hard, so let's iterate. Maybe something like "can-join-newer-cluster-version"?

File: etcdserver/server.go
Comment (serathius) [inline etcdserver/server.go:382]: `cfg.AcceptNextStorageVersion` should never change during runtime. Setting it post construction has a risk that we miss it. Could we pass it in `RaftCluster` constructor?

File: mvcc/downgrade.go
Comment (serathius) [inline mvcc/downgrade.go:127]: Why distinguish between confstate set and empty? Don't think the code on main branch looks the same. We should provide more context in comments

**Comment by serathius** [comment]
High level looks good, left couple of comments. Some issues possibly come from borrowing code on main branch which was written under different assumption or was just written incorrectly by me :(. Please make sure we fix those issues also on main branch.

File: embed/config.go
Comment (siyuanfoundation) [reply embed/config.go:367]: 5 words is a mouthful. How about `forward-compatible`?

**Comment by serathius** [comment]
ping @ahrtr

File: mvcc/kvstore.go
Comment (ahrtr) [inline mvcc/kvstore.go:393]: Suggest to gate this by the option `--next-cluster-version-compatible` as well.

File: etcdmain/config.go
Comment (ahrtr) [inline etcdmain/config.go:197]:
```suggestion
	fs.BoolVar(&cfg.ec.NextClusterVersionCompatible, "next-cluster-version-compatible", false, "Enable 3.4 to be compatible with next version 3.5, to allow 3.4 server to join 3.5 cluster and start on 3.5 schema")
```

File: mvcc/kvstore.go
Comment (siyuanfoundation) [reply mvcc/kvstore.go:393]: the problem with gating this is it might break the upgrade process. 
If the 3.4 cluster is started without `--next-cluster-version-compatible`, when some node is upgraded to 3.5, and some 3.4 node is falling behind. When the 3.4 node receives a snapshot from 3.5 node, we do want to do `downgradeMetaBucket` to avoid any data problems when that node upgrade to 3.5. So this step is needed regardless of `--next-cluster-version-compatible`

Comment (ahrtr) [reply mvcc/kvstore.go:393]: I understood your point.

- This is a generic problem. I would suggest to fix it in main firstly, then 3.5 and 3.4.
- I think eventually we do need to gate this under a flag just in case it has any regression.

It can be a followup.

File: etcdserver/apply.go
Comment (lavacat) [inline etcdserver/apply.go:148]: My understanding is that we add AuthStatus only for compatibility, not back-porting full feature. https://github.com/etcd-io/etcd/pull/11536/files#diff-583048be54c04f9555a60191ec8767757fc7f0c477d5bc7f2191af8a3c1bf920

In this case, maybe it's simpler to `return nil` here and not backport applyV3.AuthStatus() and AuthStatusResponse.

File: etcdserver/verify/verify.go
Comment (lavacat) [inline etcdserver/verify/verify.go:78]: This looks like a different implementation of `Verify` from main and `release-3.5` branch.
https://github.com/etcd-io/etcd/blob/main/server/verify/verify.go#L51
It's somewhat confusing for code readability across branches. And we are adding all this code to just check `confstate` and `term` buckets. 
Is this feature required? Looks like `nice-to-have`.

File: mvcc/kvstore.go
Comment (siyuanfoundation) [reply mvcc/kvstore.go:394]: added flag check.

File: etcdserver/verify/verify.go
Comment (siyuanfoundation) [reply etcdserver/verify/verify.go:78]: This is to ensure we did not miss any part that let the 3.5 schema through. Adding this is more ensuring for supporting something that 3.4 does not support before. The feature is mentioned in https://docs.google.com/document/d/1siSo-tA_xLRo5li5wK8KqWsOiq34PIgtmuSDcyxxwsY/edit#heading=h.efr9asc472y. 

In theory, this schema verification should be in 3.5 and main as well once downgrade feature is complete. I don't see this as different implementation that confuses readability.

File: etcdserver/apply.go
Comment (siyuanfoundation) [reply etcdserver/apply.go:148]: This can be optional. But I think having the real response is
1. better for readability. 
2. 3.4 does not really know what the response could be used for in 3.5. It is probably safer to return the real response.

File: mvcc/downgrade.go
Comment (siyuanfoundation) [reply mvcc/downgrade.go:112]: Added a comment. As for test, I could not find any difference in the data schemas of 3.3 and 3.4, not sure about what to test.

File: etcdserver/verify/verify.go
Comment (lavacat) [reply etcdserver/verify/verify.go:78]: 1. I think original motivation for [Verify](https://github.com/etcd-io/etcd/commit/47b28b600adf210fe8e63c5484f63abea16d84fb) was ConsistencyIndex bug that was fixed later. IMO it's using env var as a guard because validateWal is expensive. There were no significant changes after original commit.
2. Your backport has completely different logic. Also removes `ExactIndex` in config.
3. At the moment there is no schema verification in 3.5/main Verify. But there is [schema.Validate](https://github.com/etcd-io/etcd/blob/main/server/etcdserver/bootstrap.go#L239) as part of NewServer logic. So, I'd add your validateSchema in server.NewServer after backend is initialized.

@ahrtr @serathius Please make a final call on this.

**Comment by ahrtr** [comment]
Overall looks good. We need to add a couple of e2e test a well,
- The normal upgrade from 3.3 to 3.4. We need to ensure the normal upgrading from 3.3 to 3.4 isn't impacted by this PR.
- Downgrade from 3.5 to 3.4 
    - `next-cluster-version-compatible = false`
    - `next-cluster-version-compatible = true`

It can be addressed in a followup, but please manually verify the cases.

File: etcdserver/verify/verify.go
Comment (ahrtr) [reply etcdserver/verify/verify.go:78]: Ideally, the best approach should be always implement in `main` firstly, and then backport to stable releases (3.5 and then 3.4). For this case, we should ideally
- Backport the [verification](https://github.com/etcd-io/etcd/blob/main/server/verify/verify.go) from `main` to 3.4 (already backported to 3.5); This can be a separate PR. Note that we are adding more and more changes in 3.4, this is definitely nice to have to add more protection & verification.
- Add the `validateSchema` in `main`, then backport to 3.5 and 3.4.  This can be discussed separately. (The `schema.Validate` is a different story, it was added in https://github.com/etcd-io/etcd/pull/13200 and never be backported to 3.5). 

For now, I don't want to block this PR as long as they can be addressed in followup PRs. Also https://github.com/etcd-io/etcd/pull/17330#issuecomment-1944436289

File: etcdserver/apply.go
Comment (lavacat) [reply etcdserver/apply.go:148]: unless you add [EtcdServer.AuthStatus ](https://github.com/etcd-io/etcd/pull/11536/files#diff-583048be54c04f9555a60191ec8767757fc7f0c477d5bc7f2191af8a3c1bf920) that registers request, response won't be used anywhere. This [Trigger](https://github.com/etcd-io/etcd/blob/release-3.4/etcdserver/server.go#L2275) in apply will be noop.

Comment (siyuanfoundation) [reply etcdserver/apply.go]: set to nil

**Comment by siyuanfoundation** [comment]
Tested 3.3 -> 3.4 upgrade manually with:

1. start 3.3 cluster
```
../etcd-3.3/etcd --name infra1 --listen-client-urls http://127.0.0.1:2379 --advertise-client-urls http://127.0.0.1:2379 --listen-peer-urls http://127.0.0.1:12380 --initial-advertise-peer-urls http://127.0.0.1:12380 --initial-cluster-token etcd-cluster-1 --initial-cluster 'infra1=http://127.0.0.1:12380,infra2=http://127.0.0.1:22380,infra3=http://127.0.0.1:32380' --initial-cluster-state new

../etcd-3.3/etcd --name infra2 --listen-client-urls http://127.0.0.1:22379 --advertise-client-urls http://127.0.0.1:22379 --listen-peer-urls http://127.0.0.1:22380 --initial-advertise-peer-urls http://127.0.0.1:22380 --initial-cluster-token etcd-cluster-1 --initial-cluster 'infra1=http://127.0.0.1:12380,infra2=http://127.0.0.1:22380,infra3=http://127.0.0.1:32380' --initial-cluster-state new

../etcd-3.3/etcd --name infra3 --listen-client-urls http://127.0.0.1:32379 --advertise-client-urls http://127.0.0.1:32379 --listen-peer-urls http://127.0.0.1:32380 --initial-advertise-peer-urls http://127.0.0.1:32380 --initial-cluster-token etcd-cluster-1 --initial-cluster 'infra1=http://127.0.0.1:12380,infra2=http://127.0.0.1:22380,infra3=http://127.0.0.1:32380' --initial-cluster-state new
```
2. verify status
```
go run tools/benchmark/main.go put --total=10000000 --key-size=64 --val-size=512`

bin/etcdctl --endpoints=http://127.0.0.1:2379,http://127.0.0.1:22379,http://127.0.0.1:32379 endpoint status -w table
```
3. upgrade server one by one
```
../etcd-3.4/bin/etcd --name infra3 --listen-client-urls http://127.0.0.1:32379 --advertise-client-urls http://127.0.0.1:32379 --listen-peer-urls http://127.0.0.1:32380 --initial-advertise-peer-urls http://127.0.0.1:32380 --initial-cluster 'infra1=http://127.0.0.1:12380,infra2=http://127.0.0.1:22380,infra3=http://127.0.0.1:32380' --initial-cluster-state existing --data-dir infra3.etcd 

../etcd-3.4/bin/etcd --name infra2 --listen-client-urls http://127.0.0.1:22379 --advertise-client-urls http://127.0.0.1:22379 --listen-peer-urls http://127.0.0.1:22380 --initial-advertise-peer-urls http://127.0.0.1:22380 --initial-cluster-token etcd-cluster-1 --initial-cluster 'infra1=http://127.0.0.1:12380,infra2=http://127.0.0.1:22380,infra3=http://127.0.0.1:32380' --initial-cluster-state existing --data-dir infra2.etcd

../etcd-3.4/bin/etcd --name infra1 --listen-client-urls http://127.0.0.1:2379 --advertise-client-urls http://127.0.0.1:2379 --listen-peer-urls http://127.0.0.1:12380 --initial-advertise-peer-urls http://127.0.0.1:12380 --initial-cluster-token etcd-cluster-1 --initial-cluster 'infra1=http://127.0.0.1:12380,infra2=http://127.0.0.1:22380,infra3=http://127.0.0.1:32380' --initial-cluster-state existing --data-dir infra1.etcd
```
4. verify status
```
bin/etcdctl --endpoints=http://127.0.0.1:2379,http://127.0.0.1:22379,http://127.0.0.1:32379 endpoint status -w table
```

File: mvcc/downgrade.go
Comment (serathius) [inline mvcc/downgrade.go:41]:
```suggestion
			zap.ByteStrings("remove-keys", [][]byte{confStateKeyName, termKeyName}),
```

**Review by serathius — APPROVED** [review]

**Comment by ahrtr** [comment]
cc @fuweid to take a look. This PR should be also important to AKS.

**Review by fuweid — COMMENTED** [review]
The change looks good. But it seems that this patch needs `go mod tidy`.

**Comment by ahrtr** [comment]
> But it seems that this patch needs `go mod tidy`.

This PR doesn't change anything for go.mod and go.sum files. I do not see any issue when running "go mod tidy" command.  Please ensure you use exactly the same go version as https://github.com/etcd-io/etcd/blob/release-3.4/.go-version#L1

**Comment by fuweid** [comment]
> This PR doesn't change anything for go.mod and go.sum files. I do not see any issue when running "go mod tidy" command.

If you pull this pull request in your local, you will see that. This pull request doesn't rebase so it shows diff.
It's fine. It's not blocking comment.

**Comment by ahrtr** [comment]
> If you pull this pull request in your local, you will see that. This pull request doesn't rebase so it shows diff.

Yes, it's true. Curious why the workflow did not fail.

Anyway, please rebase this PR. @siyuanfoundation

**Comment by siyuanfoundation** [comment]
rebased.  the workflow still passes.

**Review by ahrtr — APPROVED** [review]

**Pull request merged**
